src/supervisor/mod.rs

```diff
@@ -64,16 +71,21 @@ pub struct Spec {
     pub meta: SpecMeta,
 }
 
+type SupervisorResult = Result<
+    (Spec, Result<(), Arc<TerminationError>>), /* result when supervisor starts without errors */
+    Option<(Spec, Arc<StartError>)>,           /* result when the supervisor fails to start */
+>;
```

Comment:
Whoah, this is an intense type!

Most of this makes sense, but I'm a bit confused by the `Option` as the Err type. What does it mean to have an `Ok(None)`?

I would rather see this look like a single `Result` with an enum allowing the possibility of either an `Arc<TerminationError>` or an `Arc<StartError>`. And then maybe the Spec can be returned unconditionally alongside the `Result`. Does that make sense?  Or rather, maybe an error enum that can be inside the `Arc`?

Reply:
> Whoah, this is an intense type!

Yes, when we need comments for it, this is a tail sign.

> What does it mean to have an `Ok(None)`

You would not have an `Ok(None)` here, a few description on what this mean: 

* If it is `Ok`, it means it did not fail at start
* If it is `Ok((spec, Err(_))` it means we finished with a termination error
* If it is `Ok((spec, Ok(_))` it means we finished with no termination error

There are many combinations here because this functions serves for both sync and async interactions, the complexity is inherent and the type value here makes it obvious.

Now that I've been away from the computer, I think it make more sense to make this a custom enum type, I'll try that